controller/contour/deployment.go

```diff
@@ -0,0 +1,423 @@
+	stringsutil "github.com/projectcontour/contour-operator/util/strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	// deployContainerName is the name of the Contour container.
+	deployContainerName = "contour"
+	// deployImageName is the image name of the Contour container.
+	deployImageName = "docker.io/projectcontour/contour:main"
+	// deployCfgMapName is the name of the Contour configmap.
+	deployCfgMapName = "contour"
+	// deployNsEnvVar is the name of the contour namespace environment variable.
+	deployNsEnvVar = "CONTOUR_NAMESPACE"
+	// deployPodEnvVar is the name of the contour pod name environment variable.
+	deployPodEnvVar = "POD_NAME"
+	// certsVolumeName is the name of the contour certificates volume.
+	certsVolumeName = "contourcert"
+	// certsVolumeMountDir is the directory name of the contour certificates volume.
+	certsVolumeMountDir = "certs"
+	// certsSecretName is the name of the secret used as the certificate volume source.
+	certsSecretName = certsVolumeName
+	// cfgVolumeName is the name of the contour configuration volume.
+	cfgVolumeName = "contour-config"
+	// cfgVolumeMountDir is the directory name of the contour configuration volume.
+	cfgVolumeMountDir = "config"
+	// cfgFileName is the name of the contour configuration file.
+	cfgFileName = "contour.yaml"
+)
+
+// ensureDeployment ensures a deployment exists for the given contour.
+func (r *Reconciler) ensureDeployment(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	haveDepl, current, err := r.currentDeployment(ctx, contour)
+	if err != nil {
+		return err
+	}
+
+	version, err := stringsutil.VersionFromImage(deployImageName)
+	if err != nil {
+		return fmt.Errorf("failed to parse version from image %s: %v", deployImageName, err)
+	}
+	desired, err := desiredDeployment(contour, version)
+	if err != nil {
+		return fmt.Errorf("failed to build deployment: %v", err)
+	}
+
+	switch {
+	case !haveDepl:
+		if err := r.createDeployment(ctx, desired); err != nil {
+			return err
+		}
+	case haveDepl:
+		if _, err := r.updateDeployment(ctx, current, desired); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ensureDeploymentDeleted ensures the deployment for the provided contour
+// is deleted.
+func (r *Reconciler) ensureDeploymentDeleted(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	key := types.NamespacedName{
+		Namespace: contour.Spec.Namespace.Name,
+		Name:      contour.Name,
+	}
+
+	deployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: key.Namespace,
+			Name:      key.Name,
+		},
+	}
+
+	if err := r.Client.Delete(ctx, deployment); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		return nil
```

Comment:
Why return nil rather than the deletion error?

Reply:
`ensureDeploymentDeleted()` is ensuring the deployment for the provided contour is deleted. If the error returned by the client delete operation is `NotFound` then the deployment is already deleted.

Reply:
Oh I misread it; I was expecting
```Go
		if errors.IsNotFound(err) {
			return nil
		}

		return err
```